File: Problem3.java
// Time Complexity : O(n^2)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No
class Solution {
    public int depthSum(List<NestedInteger> nestedList) {
        return dfs(nestedList,1);
    }
    public int dfs(List<NestedInteger> nestedList, int depth){
        int result = 0;
        for(NestedInteger el : nestedList){
            if(el.isInteger()){
                result += depth*el.getInteger();
            }
            else{
                result += dfs(el.getList(),depth+1);
            }
        }
        return result;
    }
}

File: Problem_3.cpp

339. Nested List Weight Sum

TC O(n)
SC O(n)

class Solution {
public:
    int depthSum(vector<NestedInteger>& nestedList) {
        queue<NestedInteger> q;
        for (NestedInteger nested : nestedList) {
            q.push(nested);
        }

        int depth = 1;
        int total = 0;

        while (!q.empty()) {
            size_t size = q.size();
            for (size_t i = 0; i < size; i++) {
                NestedInteger nested = q.front();
                q.pop();
                if (nested.isInteger()) {
                    total += nested.getInteger() * depth;
                } else {
                    for (NestedInteger nested_deeper : nested.getList()) {
                        q.push(nested_deeper);
                    }
                }
            }
            depth++;
        }
        return total;
    }
};
